### DISCUSSION TITLE
How to add custom jinja2 filters

### DISCUSSION THREAD

**Discussion by KomodoGal** [GhostManager/Ghostwriter#284] [Q&A] [answered]
Hi :)

We're really enjoying using Ghostwriter for our reports but felt that inline comments were a missing feature. I think this is a relatively easy problem to solve with a filter. Each comment begins and ends with @@:
```
def strip_comments(paragraph):
    splitPara = paragraph.split("@@")
    if len(splitPara) % 2 == 0:
        raise InvalidFilterValue(
            "You probably forgot to close a comment :/ please make sure to start and end with @@"
        )
    else:
        splitPara = [para.strip() for para in splitPara]
        return splitPara[::2].join(' ')

```
I'm not sure it works, this is because I haven't been able to use it in our documents. I added the function and a line in the "prepare_jinja2_env" function to include it (in Ghostwriter/ghostwriter/modules/reportwriter.py), but I don't see in the filters using {% debug %}.
I tried restarting the containers but there was no difference.

How do you add custom code and filters?

**Comment by KomodoGal** [comment, marked as answer]
I looked at [https://www.ghostwriter.wiki/development/modifying-code](https://www.ghostwriter.wiki/development/modifying-code) in the wiki and found my answer. 
After changing code run:
docker-compose -f production.yml stop; docker-compose -f production.yml rm -f; docker-compose -f production.yml build; docker-compose -f production.yml up -d

**Comment by KomodoGal** [comment]
By the way, the correct filter is:
```
def strip_comments(paragraph):
    splitPara = paragraph.split("@@")
    if len(splitPara) % 2 == 0:
        raise InvalidFilterValue(
            "You probably forgot to close a comment :/ please make sure to start and end with @@"
        )
    else:
        splitPara = [para.strip() for para in splitPara]
        return ' '.join(splitPara[::2])
```

and don't forget to add:
```
# custom filters :)
env.filters["strip_comments"] = strip_comments
```
to the function prepare_jinja2_env in ghostwriter/modules/reportwriter.py

In the document use strip_html first

**Reply by KomodoGal** [reply]
A better function is:
```
def strip_comments(text):
    text = re.sub('\n@@.*?@@\n', '\n', text)
    text = re.sub('^@@.*?@@\n', '', text)
    text = re.sub('\n@@.*?@@$', '', text)
    return re.sub('@@.*?@@ |@@.*?@@', '', text)
```
To have it run on RichText include this line in the function _replace_and_write in ghostwriter/modules/reportwriter.py:
`text = strip_comments(text)`

put it right after `text = text.replace("\r\n", "")`

then remember to run:
docker-compose -f production.yml stop; docker-compose -f production.yml rm -f; docker-compose -f production.yml build; docker-compose -f production.yml up -d

**Comment by chrismaddalena** [comment]
Hey @KomodoGuy, thanks for starting this discussion and updating it with your solution! Support for inline comments is a neat idea I can see implementing. I'll add it to the list for consideration for the road map.
